File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/utilities/ghostboxutility/GhostBoxFragment.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.utilities.ghostboxutility;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.media.audiofx.Visualizer;
import android.os.Bundle;
import android.speech.tts.TextToSpeech;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.ArrayRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.InvestigationFragment;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.utilities.ghostboxutility.data.GhostBoxUtilityModel;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.utilities.ghostboxutility.views.WaveformView;
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.PermissionsViewModel;

import java.util.Arrays;
import java.util.Locale;

/**
 * ToolGhostSpeakFragment class
 *
 * @author TritiumGamingStudios
 */
public class GhostBoxFragment extends InvestigationFragment
        implements Visualizer.OnDataCaptureListener {

    private PermissionsViewModel permissionsViewModel;

    @Nullable
    private WaveformView waveFormView = null;
    @Nullable
    private TextToSpeech textToSpeech = null;
    @Nullable
    private Visualizer visualizer = null;

    public GhostBoxFragment() {
        super(R.layout.fragment_utilities_ghostbox);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @SuppressLint({"UseCompatLoadingForDrawables", "ResourceType"})
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (permissionsViewModel == null) {
            permissionsViewModel = new ViewModelProvider(
                    requireActivity()).get(PermissionsViewModel.class);
        }

        AppCompatTextView title = view.findViewById(R.id.textView_title);
        AppCompatImageView prev = view.findViewById(R.id.button_prev);

        LinearLayout scrollview_list1 = view.findViewById(R.id.linearlayout_scrollview_list1);
        LinearLayout scrollview_list2 = view.findViewById(R.id.linearlayout_scrollview_list2);
        LinearLayout scrollview_list3 = view.findViewById(R.id.linearlayout_scrollview_list3);

        waveFormView = view.findViewById(R.id.waveFormView);


        title.setText("Ghost Box");


        prev.setOnClickListener(v ->
                Navigation.findNavController(v).popBackStack());

        if (getContext() != null) {
            setScrollListEntries(
                    getContext(), view, R.array.ghostspeaktool_general_array, scrollview_list1);
            setScrollListEntries(
                    getContext(), view, R.array.ghostspeaktool_spiritbox_array, scrollview_list2);
            setScrollListEntries(
                    getContext(), view, R.array.ghostspeaktool_oijaboard_array, scrollview_list3);

            if (!permissionsViewModel.isRecordAudioAllowed()) {
                try {
                    requestAudioPermissions(requireContext(), requireActivity());
                } catch (IllegalStateException e) {
                    e.printStackTrace();
                }
            }
        }

        waveFormView.updateVisualizer(null);

        startTextToSpeech();

        try {
            requireActivity()
                    .getOnBackPressedDispatcher()
                    .addCallback(getViewLifecycleOwner(),
                    new OnBackPressedCallback(true) {
                        @Override
                        public void handleOnBackPressed() {
                            Navigation.findNavController(view).popBackStack();
                        }
                    });
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    /**
     * @param context     The context
     * @param arrayRes    The Resource array of Strings holding phrases
     * @param destination The target LinearLayout
     */
    public void setScrollListEntries(@NonNull Context context,
                                     View view,
                                     @ArrayRes int arrayRes,
                                     @NonNull LinearLayout destination) {
        LayoutInflater inflater;
        try {
            inflater = LayoutInflater.from(requireContext());
        } catch (IllegalStateException e) {
            e.printStackTrace();
            return;
        }
        GhostBoxUtilityModel toolGhostSpeakData = new GhostBoxUtilityModel();

        int[] spiritBoxEntries = toolGhostSpeakData.getEntries(context, arrayRes);
        for (int e : spiritBoxEntries) {
            View evidenceParent = inflater.inflate(
                    R.layout.item_ghostbox_entry,
                    (ViewGroup) view,
                    false);

            AppCompatTextView text = evidenceParent.findViewById(R.id.textView_entry);

            text.setText(getResources().getString(e));

            text.setOnClickListener(v -> new Thread(() -> {
                startTextToSpeech();
                if (textToSpeech != null && permissionsViewModel.isRecordAudioAllowed()) {
                    startVisualizer();
                    textToSpeech.speak(
                            text.getText(), TextToSpeech.QUEUE_FLUSH, null, null);
                } else {
                    Log.d("TTS",
                            "TTS is null, or RECORD_AUDIO is " +
                                    permissionsViewModel.isRecordAudioAllowed());
                }

            }).start());

            destination.addView(evidenceParent);
        }
    }

    /**
     *
     */
    private void startVisualizer() {
        if (visualizer == null) {
            visualizer = new Visualizer(0);
            visualizer.setDataCaptureListener(
                    this, Visualizer.getMaxCaptureRate(), true, false);
            int CAPTURE_SIZE = 44100;
            visualizer.setCaptureSize(CAPTURE_SIZE);
            visualizer.setEnabled(true);
        }
    }

    /**
     *
     */
    private void stopVisualizer() {
        if (visualizer != null) {
            visualizer.setEnabled(false);
            visualizer.release();
            visualizer = null;
        }
    }

    /**
     *
     */
    private void startTextToSpeech() {
        if (textToSpeech == null) {
            try {
                textToSpeech = new TextToSpeech(requireContext(), status -> {
                    if (status == TextToSpeech.SUCCESS) {
                        int result = textToSpeech.setLanguage(Locale.getDefault());
                        if (result == TextToSpeech.LANG_MISSING_DATA ||
                                result == TextToSpeech.LANG_NOT_SUPPORTED) {
                            Log.e("TTS",
                                    "Language " +
                                            Locale.getDefault().getDisplayCountry() + " not supported");
                            stopTextToSpeech();
                            visitTTSVoiceDownloads();
                        }
                    } else {
                        Log.d("TTS", "Initialization failed");
                        stopTextToSpeech();
                        visitTTSVoiceDownloads();
                    }
                });
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     *
     */
    private void stopTextToSpeech() {
        if (textToSpeech != null) {
            textToSpeech.stop();
            textToSpeech.shutdown();
            textToSpeech = null;
        }
    }

    /**
     * visitTTSVoiceDownloads method
     * <p>
     * The solution was to turn on Wi-Fi on the device and add German and Russian in
     * "Settings -> Language & Input -> Google voice typing -> Voices".
     * After that the languages were downloaded and the app worked as desired.
     */
    public void visitTTSVoiceDownloads() {
        try {
            Intent intent = new Intent(TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            try {
                requireContext().startActivity(intent);
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        } catch (Exception e) {
            try {
                requireActivity().runOnUiThread(() -> {
                    Toast toast = Toast.makeText(getContext(),
                            "Could not locate Text to Speech Settings Page", Toast.LENGTH_LONG);
                    toast.show();
                });
            } catch (IllegalStateException e2) {
                e2.printStackTrace();
            }
            e.printStackTrace();
        }
    }

    /**
     * @param thisVisualiser The visualizer
     * @param waveform       The waveform
     * @param samplingRate   The sample rate
     */
    @Override
    public void onWaveFormDataCapture(Visualizer thisVisualiser,
                                      @NonNull byte[] waveform,
                                      int samplingRate) {
        if (waveFormView != null) {
            if (textToSpeech != null && textToSpeech.isSpeaking()) {
                waveFormView.updateVisualizer(Arrays.copyOf(waveform, waveform.length));
            } else {
                waveFormView.updateVisualizer(null);
            }
        }
    }

    /**
     * @param thisVisualiser The visualizer
     * @param fft            The fft
     * @param samplingRate   The sampling rate
     */
    @Override
    public void onFftDataCapture(Visualizer thisVisualiser, byte[] fft, int samplingRate) {
        // NO-OP
    }

    /**
     * requestAudioPermissions
     * <p>
     * Allows the user to allow or deny RECORD_AUDIO system function. This will restrict
     * the Sprit Box utility at its most fundamental level.
     *
     * @param c The context
     * @param a The current activity
     */
    private void requestAudioPermissions(@NonNull Context c, @NonNull Activity a) {

        if (ContextCompat.checkSelfPermission(
                c, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {

            // When permission is not granted by user,
            // show them message why this permission is needed.
            int MY_PERMISSIONS_RECORD_AUDIO = 1;
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    a,
                    Manifest.permission.RECORD_AUDIO)) {
                //Give user option to still opt-in the permissions
                ActivityCompat.requestPermissions(
                        a,
                        new String[]{Manifest.permission.RECORD_AUDIO},
                        MY_PERMISSIONS_RECORD_AUDIO);

                Log.d("RecordAudio", "Accepted1");
            } else {
                // Show user dialog to grant permission to record audio
                ActivityCompat.requestPermissions(
                        a,
                        new String[]{Manifest.permission.RECORD_AUDIO},
                        MY_PERMISSIONS_RECORD_AUDIO);

                Log.d("RecordAudio", "Accepted2");
            }
        }

        // If permission is granted, then go ahead recording audio
        if (ContextCompat.checkSelfPermission(c,
                Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {

            permissionsViewModel.setRecordAudioAllowed(true);

            Log.d("RecordAudio", "Accepted3");
        }
    }

    @Override
    protected void initViewModels() {
        super.initViewModels();
    }

    @Override
    public void reset() { }

    /**
     *
     */
    @Override
    public void onDestroy() {

        stopTextToSpeech();
        stopVisualizer();

        super.onDestroy();
    }

}

File: app-compose/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/ui/noncompose/investigation/utilities/ghostboxutility/GhostBoxFragment.kt
package com.tritiumgaming.phasmophobiaevidencepicker.ui.noncompose.investigation.utilities.ghostboxutility

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.audiofx.Visualizer
import android.media.audiofx.Visualizer.OnDataCaptureListener
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.annotation.ArrayRes
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.navigation.Navigation.findNavController
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.data.model.codex.ghostbox.GhostBoxUtilityModel
import com.tritiumgaming.phasmophobiaevidencepicker.ui.noncompose.investigation.InvestigationFragment
import com.tritiumgaming.phasmophobiaevidencepicker.ui.noncompose.investigation.utilities.ghostboxutility.views.WaveformView
import java.util.Locale

class GhostBoxFragment :
    InvestigationFragment(R.layout.fragment_utilities_ghostbox), OnDataCaptureListener {
    private var waveFormView: WaveformView? = null
    private var textToSpeech: TextToSpeech? = null
    private var visualizer: Visualizer? = null

    @SuppressLint("UseCompatLoadingForDrawables", "ResourceType")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        val title = view.findViewById<AppCompatTextView>(R.id.textView_title)
        val prev = view.findViewById<AppCompatImageView>(R.id.button_left)
        val promptList1 = view.findViewById<LinearLayout>(R.id.linearlayout_scrollview_list1)
        val promptList2 = view.findViewById<LinearLayout>(R.id.linearlayout_scrollview_list2)
        val promptList3 = view.findViewById<LinearLayout>(R.id.linearlayout_scrollview_list3)
        waveFormView = view.findViewById(R.id.waveFormView)

        title?.text = getString(R.string.ghostbox_title)

        prev.setOnClickListener { v: View? -> v?.let { findNavController(v).popBackStack() } }

        setScrollListEntries(
            requireContext(), view, R.array.ghostspeaktool_general_array, promptList1)
        setScrollListEntries(
            requireContext(), view, R.array.ghostspeaktool_spiritbox_array, promptList2)
        setScrollListEntries(
            requireContext(), view, R.array.ghostspeaktool_oijaboard_array, promptList3)

        if (!permissionsViewModel.isRecordAudioAllowed) {
            try { requestAudioPermissions(requireContext(), requireActivity()) }
            catch (e: IllegalStateException) { e.printStackTrace() }
        }

        waveFormView?.updateVisualizer(null)

        startTextToSpeech()

        try {
            requireActivity().onBackPressedDispatcher.addCallback(
                viewLifecycleOwner,
                object : OnBackPressedCallback(true) {
                    override fun handleOnBackPressed() {
                        findNavController(view).popBackStack()
                    }
                })
        } catch (e: IllegalStateException) { e.printStackTrace() }
    }

    private fun setScrollListEntries(
        context: Context, view: View?, @ArrayRes arrayRes: Int, destination: LinearLayout) {
        val inflater: LayoutInflater
        try { inflater = LayoutInflater.from(requireContext())
        } catch (e: IllegalStateException) { e.printStackTrace()
            return }
        val toolGhostSpeakData = GhostBoxUtilityModel()

        val spiritBoxEntries = toolGhostSpeakData.getEntries(context, arrayRes)
        for (e in spiritBoxEntries) {
            val evidenceParent = inflater.inflate(
                R.layout.item_ghostbox_entry, view as ViewGroup?, false)

            val text = evidenceParent.findViewById<AppCompatTextView>(R.id.textView_entry)

            text.text = resources.getString(e)

            text.setOnClickListener {
                Thread {
                    permissionsViewModel.let { permissionsViewModel ->
                        startTextToSpeech()
                        if (permissionsViewModel.isRecordAudioAllowed) {
                            startVisualizer()
                            textToSpeech?.speak(
                                text.text, TextToSpeech.QUEUE_FLUSH, null, null) }
                        else {
                            Log.d("TTS",
                                "TTS is null, or RECORD_AUDIO is " +
                                        "${permissionsViewModel.isRecordAudioAllowed}") }
                    }
                }.start()
            }

            destination.addView(evidenceParent)
        }
    }

    private fun startVisualizer() {
        visualizer = visualizer ?: Visualizer(0)
        visualizer?.let { visualizer ->
            visualizer.setDataCaptureListener(
                this, Visualizer.getMaxCaptureRate(), true, false)
            val CAPTURE_SIZE = 44100
            visualizer.setCaptureSize(CAPTURE_SIZE)
            visualizer.setEnabled(true)
        }
    }

    private fun stopVisualizer() {
        visualizer?.let { visualizer ->
            visualizer.setEnabled(false)
            visualizer.release()
        }
        visualizer = null
    }

    private fun startTextToSpeech() {
        if (textToSpeech == null) {
            try {
                textToSpeech = TextToSpeech(requireContext()) { status: Int ->
                    if (status == TextToSpeech.SUCCESS) {
                        val result = textToSpeech!!.setLanguage(Locale.getDefault())
                        if (result == TextToSpeech.LANG_MISSING_DATA ||
                            result == TextToSpeech.LANG_NOT_SUPPORTED
                        ) {
                            Log.e(
                                "TTS",
                                "Language " +
                                        Locale.getDefault().displayCountry + " not supported"
                            )
                            stopTextToSpeech()
                            visitTTSVoiceDownloads()
                        }
                    } else {
                        Log.d("TTS", "Initialization failed")
                        stopTextToSpeech()
                        visitTTSVoiceDownloads()
                    }
                }
            } catch (e: IllegalStateException) { e.printStackTrace() }
        }
    }

    private fun stopTextToSpeech() {
       textToSpeech?.let { textToSpeech ->
           textToSpeech.stop()
           textToSpeech.shutdown()
       }
        textToSpeech = null
    }

    /**
     * The solution was to turn on Wi-Fi on the device and add German and Russian in
     * "Settings -> Language & Input -> Google voice typing -> Voices".
     * After that the languages were downloaded and the app worked as desired.
     */
    private fun visitTTSVoiceDownloads() {
        try {
            val intent = Intent(TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            try { requireContext().startActivity(intent) }
            catch (e: IllegalStateException) { e.printStackTrace() }
        } catch (e: Exception) {
            try {
                requireActivity().runOnUiThread {
                    val toast = Toast.makeText(
                        context,
                        "Could not locate Text to Speech Settings Page", Toast.LENGTH_LONG
                    )
                    toast.show()
                }
            } catch (e2: IllegalStateException) { e2.printStackTrace() }
            e.printStackTrace()
        }
    }

    /**
     * @param thisVisualiser The visualizer
     * @param waveform       The waveform
     * @param samplingRate   The sample rate
     */
    override fun onWaveFormDataCapture(
        thisVisualiser: Visualizer,
        waveform: ByteArray,
        samplingRate: Int
    ) {
        waveFormView?.let { waveFormView ->
            if (textToSpeech?.isSpeaking == true) {
                waveFormView.updateVisualizer(waveform.copyOf(waveform.size)) }
            else { waveFormView.updateVisualizer(null) }
        }
    }

    /**
     * @param thisVisualiser The visualizer
     * @param fft            The fft
     * @param samplingRate   The sampling rate
     */
    override fun onFftDataCapture(thisVisualiser: Visualizer, fft: ByteArray, samplingRate: Int) {
        // NO-OP
    }

    /**
     * Allows the user to allow or deny RECORD_AUDIO system function. This will restrict
     * the Sprit Box utility at its most fundamental level.
     *
     * @param c The context
     * @param a The current activity
     */
    private fun requestAudioPermissions(c: Context, a: Activity) {
        if (ContextCompat.checkSelfPermission(c, Manifest.permission.RECORD_AUDIO) !=
            PackageManager.PERMISSION_GRANTED) {

            // When permission is not granted by user,
            // show them message why this permission is needed.
            val recordAudioPermission = 1
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    a, Manifest.permission.RECORD_AUDIO)) {
                //Give user option to still opt-in the permissions
                ActivityCompat.requestPermissions(
                    a, arrayOf(Manifest.permission.RECORD_AUDIO), recordAudioPermission)

                Log.d("RecordAudio", "Accepted1")
            } else {
                // Show user dialog to grant permission to record audio
                ActivityCompat.requestPermissions(
                    a, arrayOf(Manifest.permission.RECORD_AUDIO), recordAudioPermission)

                Log.d("RecordAudio", "Accepted2")
            }
        }

        // If permission is granted, then go ahead recording audio
        if (ContextCompat.checkSelfPermission(
                c, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
            permissionsViewModel.isRecordAudioAllowed = true

            Log.d("RecordAudio", "Accepted3")
        }
    }

    override fun reset() {}

    override fun onDestroy() {
        stopTextToSpeech()
        stopVisualizer()

        super.onDestroy()
    }
}